rust/opendp/src/meas/stability.rs

```diff
@@ -0,0 +1,113 @@
+    where MI: BaseStabilityNoise<TOC> + SensitivityMetric<Distance=TOC>,
+          TIK: Eq + Hash + Clone,
+          TIC: Integer + Clone + NumCast,
+          TOC: 'static + Float + Clone + PartialOrd + SampleLaplace + NumCast {
+    fn make3(n: usize, scale: TOC, threshold: TOC) -> Fallible<Measurement<CountDomain<TIK, TIC>, CountDomain<TIK, TOC>, MI, SmoothedMaxDivergence<TOC>>> {
+        let _n: TOC = TOC::from(n).ok_or_else(|| err!(FailedCast))?;
+        let _2: TOC = TOC::from(2).ok_or_else(|| err!(FailedCast))?;
+
+        Ok(Measurement::new(
+            SizedDomain::new(MapDomain { key_domain: AllDomain::new(), value_domain: AllDomain::new() }, n),
+            SizedDomain::new(MapDomain { key_domain: AllDomain::new(), value_domain: AllDomain::new() }, n),
+            Function::new_fallible(move |data: &HashMap<TIK, TIC>| {
+                data.into_iter()
+                    .map(|(k, c_in)| {
+                        // cast the value to TOC (output count)
+                        let c_out = TOC::from(c_in.clone()).ok_or_else(|| err!(FailedCast))?;
+                        // noise output count
+                        Ok((k.clone(), MI::noise(c_out, scale, false)?))
+                    })
+                    // remove counts that fall below threshold
+                    .filter(|res| res.as_ref().map(|(_k, c)| c >= &threshold).unwrap_or(true))
+                    // fail the whole computation if any cast or noise addition failed
+                    .collect()
+            }),
+            MI::new(),
+            SmoothedMaxDivergence::new(),
+            PrivacyRelation::new_fallible(move |&d_in: &TOC, &(eps, del): &(TOC, TOC)|{
+                let ideal_scale = d_in / (eps * _n);
+                let ideal_threshold = (_2 / del).ln() * ideal_scale + _n.recip();
+                // println!("ideal: {:?}, {:?}", ideal_sigma, ideal_threshold);
+
+                if eps.is_sign_negative() || eps.is_zero() {
+                    return fallible!(FailedRelation, "cause: epsilon <= 0")
+                }
+                if eps >= _n.ln() {
+                    return fallible!(FailedRelation, "cause: epsilon >= n.ln()");
+                }
+                if del.is_sign_negative() || del.is_zero() {
+                    return fallible!(FailedRelation, "cause: delta <= 0")
+                }
+                if del >= _n.recip() {
+                    return fallible!(FailedRelation, "cause: del >= n.ln()");
+                }
+                if scale < ideal_scale {
+                    return fallible!(FailedRelation, "cause: scale < d_in / (epsilon * n)")
+                }
+                if threshold < ideal_threshold {
+                    return fallible!(FailedRelation, "cause: threshold < (2. / delta).ln() * d_in / (epsilon * n) + 1. / n");
+                }
```

Comment:
Shouldn't some of these return false rather than generating an error? Seems like we should reserve the error for programming problems, but if it's just that the relation doesn't hold, we should return false.

Reply:
After writing the initial version of the BaseStability transformation, I couldn't get the relation to pass (it can be pretty picky about distance combinations). I wanted some way for it to explain itself so I could fix it. I was trying to use the FailedRelation variant as a special variant to pass up explanations, not programmer errors. What if, on the python side, we wrap the measurement check in a python function:

```
def check_measurement(d_in, d_out, debug=False):
    if debug:
        return odp.core.measurement_check(d_in, d_out)
    else:
        try:
            return odp.core.measurement_check(d_in, d_out)
        except OdpException as err:
            if err.variant == "FailedRelation":
                return False
            else:
                raise err
```

I am open to dropping this if you disagree. A special purpose error variant that isn't really an error can be confusing.

Reply:
Ah yeah, I could see that. This is our first time exposing relations this way, so not surprising we may need some extra info. Just feels a little funny folding it into the general Error. I'm fine keeping this for now. If it turns out to be useful to have these distinctions, then probably the right thing is to have relations return a dedicated enum type.